pkg/operator/ceph/object/controller.go

```diff
@@ -430,16 +435,38 @@ func (r *ReconcileCephObjectStore) reconcile(request reconcile.Request) (reconci
 				*cephObjectStore,
 				opcontroller.ErrorCephUpgradingRequeue(desiredCephVersion, runningCephVersion)
 		}
-		r.clusterInfo.CephVersion = *desiredCephVersion
+		r.clusterInfo.CephVersion = *runningCephVersion
+
+		shouldRotateCephxKeys, err = keyring.ShouldRotateCephxKeys(
+			cephCluster.Spec.Security.CephX.Local, *runningCephVersion, *desiredCephVersion, cephObjectStore.Status.Cephx.Local)
+		if err != nil {
+			return reconcile.Result{}, *cephObjectStore, errors.Wrap(err, "failed to determine if cephx keys should be rotated")
+		}
+		if shouldRotateCephxKeys {
+			logger.Infof("cephx keys for CephObjectStore %q will be rotated", request.NamespacedName)
```

Comment:
Is there an update to the pod spec that will trigger the daemon restart after the rotate? I must be overlooking it.

Reply:
Great catch! That's part of the plan, and I forgot to implement that.

Reply:
A couple questions:
1. Can we just set the generation id? We know it is stable for each rotation. Or why do we need a timestamp?
2. Will an annotation update cause the pod restart? I was thinking only spec updates restart it.

Reply:
We can't use generation ID because there is a chance that a reconcile restarts in the middle. 

CephFilesystem example:
1. Reconcile begins
2. Rotate key for mds.a and update its secret
3. Restart its pod
4. Reconcile begins again before mds.b is updated
5. Rotate key (again) for mds.a and update its secret
6. Restart its pod (this won't work if we use gen ID since it hasn't changed from step 3)
7. Rotate key for mds.b and update its secret
8. Restart its pod
9. Done

Reply:
I wonder if we need to persist some temporary state (e.g. in a configmap) while a key rotation is in progress, to ensure each key is only rotated once. A configmap could be created by each daemon type to keep track if it has already been rotated for the expected generation. This allows the generation id to be set in the pod spec as the stable restart mechanism. When the rotation is completed, the state could be deleted.

Reply:
Maybe I'm still not understanding how you can use a metadata resourceVersion. Would that require us to save all of the cephx keys in secrets?

Reply:
For the ones stored in secrets, we can use the resourceVersion.

For ones not in secrets, like OSDs, I'll have to find another way. 

But this works well in my testing today for RGW, and it will be a good pattern for others to follow when implementing this for the other daemons.